lib/mayaUsd/ufe/UsdConnectionHandler.cpp

```diff
@@ -88,30 +88,96 @@ bool UsdConnectionHandler::createConnection(
+            return UsdShadeConnectableAPI::ConnectToSource(dstOutput, srcInput);
+        }
+    } else {
+        UsdShadeOutput srcOutput = srcApi.CreateOutput(srcBaseName, srcUsdAttr->usdAttributeType());
+        if (dstAttrType == UsdShadeAttributeType::Input) {
+            UsdShadeInput dstInput
+                = dstApi.CreateInput(dstBaseName, dstUsdAttr->usdAttributeType());
+            return UsdShadeConnectableAPI::ConnectToSource(dstInput, srcOutput);
+        } else {
+            UsdShadeOutput   dstOutput;
+            UsdShadeMaterial dstMaterial(dstUsdAttr->usdPrim());
+            // Special case when connecting to Material outputs:
+            if (dstMaterial
+                && (dstBaseName == UsdShadeTokens->surface || dstBaseName == UsdShadeTokens->volume
+                    || dstBaseName == UsdShadeTokens->displacement)) {
+                // Create the required output based on the type of the shader node we are trying to
+                // connect:
+                UsdShadeShader       srcShader(srcUsdAttr->usdPrim());
+                PXR_NS::SdrRegistry& registry = PXR_NS::SdrRegistry::GetInstance();
+                PXR_NS::TfToken      srcInfoId;
+                srcShader.GetIdAttr().Get(&srcInfoId);
+                PXR_NS::SdrShaderNodeConstPtr srcShaderNodeDef
+                    = registry.GetShaderNodeByIdentifier(srcInfoId);
+                if (!srcShaderNodeDef) {
+                    return false;
+                }
+                TfToken renderContext = srcShaderNodeDef->GetSourceType() == "glslfx"
+                    ? UsdShadeTokens->universalRenderContext
+                    : srcShaderNodeDef->GetSourceType();
+                if (dstBaseName == UsdShadeTokens->surface) {
+                    dstOutput = dstMaterial.CreateSurfaceOutput(renderContext);
+                } else if (dstBaseName == UsdShadeTokens->volume) {
+                    dstOutput = dstMaterial.CreateVolumeOutput(renderContext);
+                } else {
+                    dstOutput = dstMaterial.CreateDisplacementOutput(renderContext);
+                }
+            } else {
+                dstOutput = dstApi.CreateOutput(dstBaseName, dstUsdAttr->usdAttributeType());
+            }
+            return UsdShadeConnectableAPI::ConnectToSource(dstOutput, srcOutput);
+        }
+    }
+
+    // Return a failure.
     return false;
 }
 
 bool UsdConnectionHandler::deleteConnection(
     const Ufe::Attribute::Ptr& srcAttr,
     const Ufe::Attribute::Ptr& dstAttr) const
 {
-    PXR_NS::UsdAttribute srcUsdAttr = usdAttrFromUfeAttr(srcAttr);
-    PXR_NS::UsdAttribute dstUsdAttr = usdAttrFromUfeAttr(dstAttr);
+    UsdAttribute* srcUsdAttr = usdAttrFromUfeAttr(srcAttr);
+    UsdAttribute* dstUsdAttr = usdAttrFromUfeAttr(dstAttr);
 
-    if (isConnected(srcUsdAttr, dstUsdAttr)) {
-        return dstUsdAttr.RemoveConnection(srcUsdAttr.GetPath());
+    if (!srcUsdAttr || !dstUsdAttr
```

Comment:
if srcUsdAttr or dstUsdAttr is null then the code should report an error e.g. TF_RUNTIME_ERROR